agent/service_manager.go

```diff
@@ -0,0 +1,237 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/structs"
+	"golang.org/x/net/context"
+)
+
+// The ServiceManager is a layer for service registration in between the agent
+// and the local state. Any services must be registered with the ServiceManager,
+// which then maintains a long-running watch of any globally-set service or proxy
+// configuration that applies to the service in order to register the final, merged
+// service configuration locally in the agent state.
+type ServiceManager struct {
+	services map[string]*serviceConfigWatch
+	agent    *Agent
+
+	sync.Mutex
+}
+
+func NewServiceManager(agent *Agent) *ServiceManager {
+	return &ServiceManager{
+		services: make(map[string]*serviceConfigWatch),
+		agent:    agent,
+	}
+}
+
+// AddService starts a new serviceConfigWatch if the service has not been registered, and
+// updates the existing registration if it has. For a new service, a call will also be made
+// to fetch the merged global defaults that apply to the service in order to compose the
+// initial registration.
+func (s *ServiceManager) AddService(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, source configSource) error {
+	reg := serviceRegistration{
+		service:  service,
+		chkTypes: chkTypes,
+		persist:  persist,
+		token:    token,
+		source:   source,
+	}
+
+	// If a service watch already exists, update the registration. Otherwise,
+	// start a new config watcher.
+	s.Lock()
+	watch, ok := s.services[service.ID]
+	s.Unlock()
+	if ok {
+		if err := watch.updateRegistration(&reg); err != nil {
+			return err
+		}
+		s.agent.logger.Printf("[DEBUG] agent.manager: updated local registration for service %q", service.ID)
+	} else {
+		// This is a new entry, so get the existing global config and do the initial
+		// registration with the merged config.
+		args := structs.ServiceConfigRequest{
+			Name:         service.Service,
+			Datacenter:   s.agent.config.Datacenter,
+			QueryOptions: structs.QueryOptions{Token: s.agent.config.ACLAgentToken},
+		}
+		if token != "" {
+			args.QueryOptions.Token = token
+		}
+		var resp structs.ServiceConfigResponse
+		if err := s.agent.RPC("ConfigEntry.ResolveServiceConfig", &args, &resp); err != nil {
```

Comment:
I don't think we should make this RPC call quite like this - it's not going through the cache which means we will end up duplicating the exact same call when the cache watch starts down below.

Initially I thought we would end up racily registering the service before the config is applied but I see the response is set directly before watch starts so the initial `serviceRegistration` event on the chan will already have the config present.

So this is correct but I think wasteful/more complex than it needs to be.

I suggest instead:

 - create watcher for service from the provided `reg`
 - call `watcher.Ready()` which will wait until the initial update is delivered and registered - note that `cache.Notify` will always deliver the initial value from the cache either immediately if it's already in cache or right after the fetch completes.
  - can have a `chan struct{}` which is made on creation of the watcher and closed when the initial config comes in (need a bool or something protected by lock to ensure we only close it the first time etc), then `ReadyWait()` can just wait on that.
  - would be kinda nice to have a timeout on this for API but not really sure how to thread that through so probably simpler to not for now and let HTTP client timeout if it would rather not wait and get an indeterminate result.

In general it's kinda interesting to consider how we deal with partial failure here. Which is worse, failing to register services locally on an agent due to being disconnected vs. registering them with out applying global defaults?

One thing to consider with this PR in general - will it cause a lot of existing setups to break in subtle ways because they now need to make a bunch of RPCs during client agent startup when services get registered but now none of those registrations will happen until the servers are available.

For file-based registrations do we even retry the registration until it succeeds or will it perma-fail if the client agent comes up before the servers are available?